website/en/docs/lang/common-mistakes.md

```diff
@@ -0,0 +1,76 @@
+Several of them are the consequence of careful tradeoffs made by the team, where you have to balance false positives and false negatives (see the above mentionned section).
+
+### Common user errors <a class="toc" id="toc-common-user-errors" href="#toc-common-user-errors"></a>
+
+#### Refinement invalidation
+
+You have to keep in mind in JavaScript, you can modify data structures as you want, and Flow takes that in consideration. As of today, Flow cannot identify side-effects, so you end up with Flow considering that a `console.log()` can modify the content of an array (since nothing prevents you from modifying `console.log`). Therefore, whenever you refine types, some of your assumptions might not be valid anymore (this is especially common within closures). This is called [refinement invalidation](https://flow.org/en/docs/lang/refinements/#toc-refinement-invalidations).
+
+#### Invalid subtyping
+
+Another difficulty people encounter is when they want to use a type different than the one declared that they believe compatible (as a parameter to a function for instance), and Flow complains, telling that the two types are not compatible.
+
+```js
+type WithError = {
+    resType: string,
+    error: Object,
+};
+type WithMaybeError = {
+    resType: string,
+    error: ?Object,
+};
+function handleAnswer(param: WithMaybeError) {}
+const res: WithError = {resType: 'withError', error: {code: 404}};
+handleAnswer(res); // Flow will complain here
+```
+
+The previous snippet of code seems correct, indeed `handleAnswer` requires a `resType`, provided by `res`, and doesn't need the error property. However, you need to consider the case where `handleAnswer` mutates its parameter.
+
+```js
+function handleAnswer(param: WithMaybeError) {
+    param.error = null;
+}
+const res: WithError = {resType: 'withError', error: {code: 404}};
+handleAnswer(res);
+console.log(res.error);
+```
+The last statement doesn't work anymore, since `res.error` has been set to null, however, `res` is a `WithError`, so it should always have an `error` field. This is the reason why Flow prevents you from doing so. To circumvent this problem, you have to indicate your [properties as covariant](https://flow.org/blog/2016/10/04/Property-Variance/) (or read-only).
+```js
+type WithError = {
+    +resType: string,
+    +error: Object,
+};
+type WithMaybeError = {
+    +resType: string,
+    +error: ?Object,
+};
+function handleAnswer(param: WithMaybeError) {}
+const res: WithError = {resType: 'withError', error: {code: 404}};
+handleAnswer(res); // Flow is happy!
+```
+
+### Errors due to Flow limitations <a class="toc" id="toc-flow-errors" href="#toc-flow-errors"></a>
```

Comment:
Can we change this section to "Why do my refinements disappear inside a closure?"

Reply:
Hmmm, I'm not sure it's the same thing. Here, it's more about "you're right, but it doesn't work with flow implementation". I could add an example with closures (which are indeed another common problem).

Reply:
For sure, I just would prefer to get the heading down to a specific question if you can. There are lots of Flow limitations and this example doesn't cover all of them :)